### PR TITLE
feat(adyen refunds): Add adyen refunds

### PR DESCRIPTION
**Pull request by ivannovosad**
## Roadmap Task

👉  https://getlago.canny.io/feature-requests/p/support-integration-with-adyen

## Context

Adyen provides payment apis for card and other payments.

## Description

Add Adyen as a payment provider supported natively

### CODE REVIEW

File: app/services/credit_notes/refunds/adyen_service.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      end
+
+      def adyen_api_key
+        organization.adyen_payment_provider.secret_key
+      end
+
+      def create_adyen_refund
+        result = client.checkout.modifications_api.refund_captured_payment(
+          adyen_refund_params,
+          payment.provider_payment_id
```
Comment (jdenquin) [inline app/services/credit_notes/refunds/adyen_service.rb:91]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/credit_notes/refunds/adyen_service.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        raise
+      end
+
+      def adyen_refund_params
+        {
+          paymentPspReference: payment.provider_payment_id,
+          merchantAccount: payment.payment_provider.merchant_account,
+          amount: {
+            value: credit_note.refund_amount_cents,
+            currency: credit_note.credit_amount_currency.upcase
```
Comment (jdenquin) [inline app/services/credit_notes/refunds/adyen_service.rb:106]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: app/services/credit_notes/refunds/adyen_service.rb [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+          }
```
Comment (jdenquin) [inline app/services/credit_notes/refunds/adyen_service.rb:107]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: app/services/payment_providers/adyen_service.rb [hunk 4]
```diff
@@ -64,6 +64,23 @@ def handle_event(organization:, event_json:)
 
         result = service.preauthorise(organization, event)
         result.raise_if_error! || result
+      when 'REFUND'
+        service = CreditNotes::Refunds::AdyenService.new
+
+        provider_refund_id = event["pspReference"]
```
Comment (jdenquin) [inline app/services/payment_providers/adyen_service.rb:70]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: app/services/payment_providers/adyen_service.rb [hunk 5]
```diff
@@ -64,6 +64,23 @@ def handle_event(organization:, event_json:)
...
+        status = event["success"] == "true" ? :succeeded : :failed
```
Comment (jdenquin) [inline app/services/payment_providers/adyen_service.rb:71]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: app/services/payment_providers/adyen_service.rb [hunk 6]
```diff
@@ -64,6 +64,23 @@ def handle_event(organization:, event_json:)
...
+
+        result = service.update_status(provider_refund_id:, status:)
+        result.raise_if_error! || result
+      when 'REFUND_FAILED'
+        return result if event["success"] != "true"
```
Comment (jdenquin) [inline app/services/payment_providers/adyen_service.rb:76]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: app/services/payment_providers/adyen_service.rb [hunk 7]
```diff
@@ -64,6 +64,23 @@ def handle_event(organization:, event_json:)
...
+
+        service = CreditNotes::Refunds::AdyenService.new
+
+        provider_refund_id = event["pspReference"]
```
Comment (jdenquin) [inline app/services/payment_providers/adyen_service.rb:80]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/factories/refund_responses.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+FactoryBot.define do
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:1]: Style/FrozenStringLiteralComment: Missing frozen string literal comment.

File: spec/factories/refund_responses.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  sequence :adyen_refunds_response do
+    OpenStruct.new(
+      response: {
+        "merchantAccount" => "#{SecureRandom.uuid}",
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:5]: Style/RedundantInterpolation: Prefer `to_s` over string interpolation.

File: spec/factories/refund_responses.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+        "pspReference" => "#{SecureRandom.uuid}",
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:6]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

Comment (jdenquin) [inline spec/factories/refund_responses.rb:6]: Style/RedundantInterpolation: Prefer `to_s` over string interpolation.

File: spec/factories/refund_responses.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+        "paymentPspReference" => "#{SecureRandom.uuid}",
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:7]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

Comment (jdenquin) [inline spec/factories/refund_responses.rb:7]: Style/RedundantInterpolation: Prefer `to_s` over string interpolation.

File: spec/factories/refund_responses.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+        "status" => "received",
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:8]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/factories/refund_responses.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+        "amount" => {
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:9]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/factories/refund_responses.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+            "currency" => "CHF",
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:10]: Layout/FirstHashElementIndentation: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

Comment (jdenquin) [inline spec/factories/refund_responses.rb:10]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/factories/refund_responses.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+            "value" => 134
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:11]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

Comment (jdenquin) [inline spec/factories/refund_responses.rb:11]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/factories/refund_responses.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+        }
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:12]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/factories/refund_responses.rb [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+      }
```
Comment (jdenquin) [inline spec/factories/refund_responses.rb:13]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/payment_providers/adyen_service_spec.rb [hunk 18]
```diff
@@ -140,5 +140,33 @@
         expect(payment_provider_service).to have_received(:preauthorise)
       end
     end
+
+    context 'when succeeded refund event' do
+      let(:refund_service) { instance_double(CreditNotes::Refunds::AdyenService) }
+
+      let(:event_json) do
+        JSON.parse(event_response_json)['notificationItems'].
```
Comment (jdenquin) [inline spec/services/payment_providers/adyen_service_spec.rb:148]: Layout/DotPosition: Place the . on the next line, together with the method name. (https://rubystyle.guide#consistent-multi-line-chains)

**Review by vincent-pochet — APPROVED** [review]

**Comment by vincent-pochet** [comment]
Looks great! :rocket:

**Pull request merged**